### ISSUE TITLE
Repeat-inside

### ISSUE THREAD

**Issue by heldr** [wix/react-templates#21]
Is there a simple way to do repeat-inside through children like [some angular plugins](https://github.com/bigblind/angular-repeat-inside) does?

Ex.:
```html
<dl repeat-inside="(word, definition) in words">
    <dt>{{word}}</dt>
    <dd>{{definition}}</dd>
</dl>
```

**Comment by eberhm** [comment]
I think you can't.
You can always do instead

```html
<dl rt-repeat="word in words">
    <dt>{word.word}</dt>
    <dd>{word.definition}</dd>
</dl>
```

**Comment by burtyish** [comment]
I ran into this question too. @eberhm, it's not quite the same cause the result will be many `dl` tags, where @heldr's question expects only one.
I found that rt-repeat forces adding extra wrapper elements for around elements in a list where they may not be wanted.

**Comment by brodybits** [comment]
Here is a possible workaround, if you don't mind an extra `<div>`:
``` html
<dl>
  <div rt-repeat="word in words">
    <dt>{word.word}</dt>
    <dd>{word.definition}</dd>
  </div>
</dl>
```

**Comment by heldr** [comment]
The idea behind **repeat-inside** is to not create an extra element

**Comment by brodybits** [comment]
> The idea behind repeat-inside is to not create an extra element

Absolutely!

The following changes to `src/reactTemplates.js` should fix the issue:

```diff
diff --git a/src/reactTemplates.js b/src/reactTemplates.js
index 47557d7..7960d0f 100644
--- a/src/reactTemplates.js
+++ b/src/reactTemplates.js
@@ -13,6 +13,7 @@ var RTCodeError = rtError.RTCodeError;
 
 var repeatTemplate = _.template('_.map(<%= collection %>,<%= repeatFunction %>.bind(<%= repeatBinds %>))');
 var ifTemplate = _.template('((<%= condition %>)?(<%= body %>):null)');
+var repeatInnerTemplate = _.template('_.map(<%= collection %>,<%= repeatInnerFunction %>.bind(<%= repeatInnerBinds %>))');
 var propsTemplateSimple = _.template('_.assign({}, <%= generatedProps %>, <%= rtProps %>)');
 var propsTemplate = _.template('mergeProps( <%= generatedProps %>, <%= rtProps %>)');
 var propsMergeFunction = 'function mergeProps(inline,external) {\n var res = _.assign({},inline,external)\nif (inline.hasOwnProperty(\'style\')) {\n res.style = _.defaults(res.style, inline.style);\n}\n' +
@@ -24,6 +25,7 @@ var simpleTagTemplate = _.template('<%= name %>(<%= props %><%= children %>)');
 var tagTemplate = _.template('<%= name %>.apply(this, [<%= props %><%= children %>])');
 var simpleTagTemplateCreateElement = _.template('React.createElement(<%= name %>,<%= props %><%= children %>)');
 var tagTemplateCreateElement = _.template('React.createElement.apply(this, [<%= name %>,<%= props %><%= children %>])');
+var tagTemplateCreateRepeatInnerElement = _.template('React.createElement.apply(this, [<%= name %>,<%= props %>,<%= innerbody %>])');
 var commentTemplate = _.template(' /* <%= data %> */ ');
 
 var templateAMDTemplate = _.template("define(<%= name ? '\"'+name + '\", ' : '' %>[<%= requirePaths %>], function (<%= requireNames %>) {\n'use strict';\n <%= injectedFunctions %>\nreturn function(){ return <%= body %>};\n});");
@@ -48,6 +50,7 @@ var templates = {
 
 var htmlSelfClosingTags = ['area', 'base', 'br', 'col', 'command', 'embed', 'hr', 'img', 'input', 'keygen', 'link', 'meta', 'param', 'source', 'track', 'wbr'];
 
+var repeatInnerProp = 'rt-repeat-inside';
 var templateProp = 'rt-repeat';
 var ifProp = 'rt-if';
 var classSetProp = 'rt-class';
@@ -306,6 +309,18 @@ function convertHtmlToReact(node, context) {
             });
         }
 
+        if (node.attribs[repeatInnerProp]) {
+            var arr = node.attribs[repeatInnerProp].split(' in ');
+            if (arr.length !== 2) {
+                throw RTCodeError.build("rt-repeat-inner invalid 'in' expression '" + node.attribs[repeatInnerProp] + "'", context, node);
+            }
+            data.item = arr[0].trim();
+            data.collection = arr[1].trim();
+            validateJS(data.item, node, context);
+            validateJS(data.collection, node, context);
+            stringUtils.addIfMissing(context.boundParams, data.item);
+            stringUtils.addIfMissing(context.boundParams, data.item + 'Index');
+        }
         if (node.attribs[templateProp]) {
             var arr = node.attribs[templateProp].split(' in ');
             if (arr.length !== 2) {
@@ -334,12 +349,21 @@ function convertHtmlToReact(node, context) {
         if (node.attribs[ifProp]) {
             data.condition = node.attribs[ifProp].trim();
         }
+
         data.children = node.children ? concatChildren(_.map(node.children, function (child) {
             var code = convertHtmlToReact(child, context);
             validateJS(code, child, context);
             return code;
         })) : '';
 
+        if (node.attribs[repeatInnerProp]) {
+            data.repeatInnerFunction = generateInjectedFunc(context, 'repeat' + stringUtils.capitalize(data.item), 'return ' + data.children.substring(1));
+            data.repeatInnerBinds = ['this'].concat(_.reject(context.boundParams, function (param) {
+                return (param === data.item || param === data.item + 'Index');
+            }));
+            data.innerbody = repeatInnerTemplate(data);
+            data.body = tagTemplateCreateRepeatInnerElement(data);
+        } else
         if (hasNonSimpleChildren(node)) {
             data.body = shouldUseCreateElement(context) ? tagTemplateCreateElement(data) : tagTemplate(data);
         } else {
```

translates the following template in `template.rt`:

```html
<dl rt-repeat="word in words">
    <dt>{word.word}</dt>
    <dd>{word.definition}</dd>
</dl>
```

to the following Javascript:

```Javascript
var templateRT = function () {
    function repeatWord1(word, wordIndex) {
        return React.createElement('dl', {}, React.createElement('dt', {}, word.word), React.createElement('dd', {}, word.definition));
    }
    return _.map(words, repeatWord1.bind(this));
};
```

Unfortunately I haven't got this working in jsfiddle yet. Will give it another try.

@heldr if this looks OK to you I am happy to issue a pull request including this change and a test to cover your scenario.

**Comment by heldr** [comment]
Wow! Gonna help u when possible :tea:

**Comment by brodybits** [comment]
Cool! Please ignore my solution above, though. Discovered it is broken.

**Comment by brodybits** [comment]
FYI here is the jsfiddle that I am using to test generated template code: http://jsfiddle.net/brodybits/85g6gebf/

This fiddle is using the generated JS from JSX using the existing `rt-repeat` tag.

**Comment by brodybits** [comment]
Sorry, I tested with the wrong code. Still testing now.

**Comment by brodybits** [comment]
@heldr I now have the fix. The problem was that in the generated inner function, I needed `[` `]` brackets around the generated child code. The last part of the diff should be changed as follows:

```diff
@@ -334,12 +349,21 @@ function convertHtmlToReact(node, context) {
         if (node.attribs[ifProp]) {
             data.condition = node.attribs[ifProp].trim();
         }
+
         data.children = node.children ? concatChildren(_.map(node.children, function (child) {
             var code = convertHtmlToReact(child, context);
             validateJS(code, child, context);
             return code;
         })) : '';
 
+        if (node.attribs[repeatInnerProp]) {
+            data.repeatInnerFunction = generateInjectedFunc(context, 'repeat' + stringUtils.capitalize(data.item), 'return [' + data.children.substring(1) + ']');
+            data.repeatInnerBinds = ['this'].concat(_.reject(context.boundParams, function (param) {
+                return (param === data.item || param === data.item + 'Index');
+            }));
+            data.innerbody = repeatInnerTemplate(data);
+            data.body = tagTemplateCreateRepeatInnerElement(data);
+        } else
         if (hasNonSimpleChildren(node)) {
             data.body = shouldUseCreateElement(context) ? tagTemplateCreateElement(data) : tagTemplate(data);
         } else {
```

http://jsfiddle.net/brodybits/6sesjtje/ shows the code generated from the following template:
```html
<dl rt-repeat-inside="word in this.state.words">
    <dt>{word.word}</dt>
    <dd>{word.definition}</dd>
</dl>
```
is working properly with the React library.

I would still like to test what happens when you combine the new `rt-repeat-inside` tag in combination with some of the other tags. So far, I know I have to make a small fix in case someone tries using `rt-repeat-inside` and `rt-repeat` in the same element. Maybe later this weekend.

**Comment by anchann** [comment]
For what it's worth, I also implemented a version of rt-repeat-inside here:

https://github.com/anchann/react-templates

**Comment by brodybits** [comment]
On Tuesday, July 28, 2015, anchann <notifications@github.com
<javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote:

> For what it's worth, I also implemented a version of rt-repeat-inside here:
>
> https://github.com/anchann/react-templates
>

rt-repeat-children, right? :+1:

>

-- 
Sent from my mobile

**Comment by bgotink** [comment]
I'd never heard of `ng-repeat-inside` before, but I don't understand why it's needed. `ng-repeat` supports a `start`/`end`, which is way more powerful:

```html
<dl>
  <dt>Word</dt>
  <dd>This is a word, we don't repeat it.</dd>
  <dt ng-repeat-start="el in list">{{ el.word }}</dt>
  <dd ng-repeat-end>{{ el.definition }}</dt>
</dl>
```

I would therefore prefer having `rt-repeat-start`/`rt-repeat-end` over `rt-repeat-inside`.

**Comment by brodybits** [comment]
> I would therefore prefer having rt-repeat-start/rt-repeat-end over rt-repeat-inside.

+1 `ng-repeat-start`/`ng-repeat-end` is also part of angular while `repeat-inside` is an Angular plugin.

**Comment by anchann** [comment]
> don't understand why needed, since ng-repeat supports...

The start/end nt-repeat directives are relatively new additions to angular, so those who have been working with angular before that time were forced to come with their own creative solutions.

> which is way more powerful

Can you share an example that can be done using start/end that cannot be done with rt-repeat-children that I've implemented? I thought about it for a bit, and couldn't come up with one.

 https://github.com/anchann/react-templates

**Comment by bgotink** [comment]
The example I've given above shows a usecase `ng-repeat-children` doesn't support. Perhaps a more clear example is this:

```html
<!-- excuse the angular -->
<div class="parent">
  <div class="child">
    This is a div containing numbers and their squares, note how this intro text isn't repeated.
  </div>
  <div ng-repeat-start="el in list" class="child">
    {{ el }}
  </div>
  <div ng-repeat-end class="child square">
    {{ el * el }}
  </div>
</div>
```

Granted, your `rt-repeat-children` supports that usecase because you discard the wrapper div, but I don't think that's a clean solution. The structures of the template and the resulting HTML code are no longer the same.

```html
<!-- template -->
<div class="parent">
  <div class="child">
    This is a div containing numbers and their squares, note how this intro text isn't repeated.
  </div>
  <div rt-repeat-inside="el in child">
    <div class="child">{ el }</div>
    <div class="child square">{ el * el }</div>
  </div>
</div>

<!-- result -->
<div class="parent">
  <div class="child">
    This is a div containing numbers and their squares, note how this intro text isn't repeated.
  </div>
  <div class="child">1</div>
  <div class="child square">1</div>
  <div class="child">2</div>
  <div class="child square">2</div>
</div>
```

This complicates things for the developer as there's no longer a clear mapping between the template and the DOM running in the end user's client. I for one don't like having elements magically disappear in the template.

**Comment by anchann** [comment]
> `rt-repeat-children` supports that

Good, so we agree that expressive power is equivalent.

> clear mapping

I agree with you that to the end user, `rt-repeat-children`'s semantics feel less clean than that of the start/end version.

From the practical standpoint, implementing start/end in the current version of react-templates would be a considerable amount of work. This is because the model works at the level of nodes. You have a node, its child content that is processed recursively, and possibly some directives, which determine the function that will wrap the generated function representation of the children. There's no notion of siblings, and no existing way to keep track of arbitrary content blocks between a pair of sibling nodes.

This practical problem brings into focus the fundamental theoretical problem with react-templates and angular style templates. We have an XML-based language describing a tree data structure. And now we want to use attributes of the XML nodes to implement control structures that transform the underlying tree. Can one choose a more creative way to set oneself up for disaster?

This problematic structure is what causes the problem with repeat-children in the first place. There is no such thing as a free-standing control flow operator, anything has to go on a node. So either you take a hit in terms of expressive power, and end up with a non-node-discarding version of rt-repeat-children, or you level-up your meta-programming, and say that to gain the desired expressive freedom, some nodes will be used purely as control flow wrappers. From this perspective, what the start/end constructs do is actually more of a hack to work around the limitations of the model, than a clean solution, while rt-repeat-children acknowledges the limitations of the given model and extends it to remove these limitations.

Coming from angular, seriously unhappy about how primitive JSX is, I originally got quite excited about react-templates. But this discussion reminded me of one point: angular's attributes-as-control-structures model is far from natural, and I only felt it was because three years of coding that way made it a habit. Angular's DOM-based templating system is what necessitated this approach, while most function-generation-based templating systems use moustache style control structures. Given that react (as far as templates are concerned) is a world of JS more so than DOM, it would have been perhaps more natural to discard the restrictions of attribute-based control structures and choose moustache style ones.

**Comment by bgotink** [comment]
I've thought about what you've written, and I cannot find a single point where I disagree. Perhaps I too have grown too accustomed to angular's directives, which try to add moustache/twig/... style logic to HTML, which is inherenly unsuited for such logic.  
It has always bothered me in angular that you simply cannot tell in which order directives are run, except by looking at the directive documentation/source code. That's because directives are applied on a DOM element (or a collection of nodes if using start/end), and to lower the number of DOM elements needed, we combine multiple directives on one element.  
Using a handlebars/twig/... approach fixes this issue by not using HTML attributes. No one can doubt what this code does:

```twig
<ul>
  <li>not repeated</li>
  {% for el in list %}
    {% if el.something %}
      <li>{{ el.something }}</li>
    {% endif %}
  {% endfor %}
</ul>
```

while the following is not clear unless you happen to know that `ng-if` has priority 500 while `ng-repeat` runs at priority 600:

```html
<ul>
  <li>not repeated</li>
  <li ng-repeat="el in list" ng-if="el.something">{{ el.something }}</li>
</ul>
```

Though it's shorter, it assumes the reader has an intimate knowledge of the used directives.  

Thanks for making me realise that this project isn't what I'm looking for :smile: I'm still not a fan of JSX, so I'll continue looking for the best way to template React.

**Comment by anchann** [comment]
Thank you for carefully thinking about my ramblings! :)

Let me know (is there a private message feature in github?) what template system you end up going with. I had a reasonably large angular app to port, and react-templates made the port fairly straight-forward, so I'm likely going to stick with it for a while. But for a new new projects, I'd be interested in trying out alternatives.

**Comment by nippur72** [comment]
would this issue be solved with the `<virtual>` tag ? (see #85)

**Comment by brodybits** [comment]
On Jan 4, 2016 8:53 AM, "nino-porcino" <notifications@github.com> wrote:
>
> would this issue be solved with the <virtual> tag ? (see #85)

I suggest this issue should be closed. It was already solved by the
rt-repeat-children tag.
